execution_engine/src/core/engine_state/upgrade.rs

```diff
@@ -161,3 +159,136 @@ impl UpgradeConfig {
         self.new_wasmless_transfer_cost
     }
 }
+
+#[derive(Clone, Error, Debug)]
+pub enum ProtocolUpgradeError {
+    #[error("Invalid upgrade config")]
+    InvalidUpgradeConfig,
+    #[error("Unable to retrieve system contract: {0}")]
+    UnableToRetrieveSystemContract(String),
+    #[error("Unable to retrieve system contract package: {0}")]
+    UnableToRetrieveSystemContractPackage(String),
+    #[error("Failed to disable previous version of system contract: {0}")]
+    FailedToDisablePreviousVersion(String),
+}
+
+pub(crate) struct UpgradeInstaller<S>
```

Comment:
Perhaps 'twould be better to say...
```suggestion
pub(crate) struct SystemUpgrader<S>
```

Reply:
did this